restoring document to a specific state using a Snapshot, #159

calibr opened the pull request:
- added public function `createDocFromSnapshot(originDoc, snapshot): Doc`
- added private function `applyDeleteItem(transaction, structs, deleteItem)` to avoid code duplication in places where delete items are needed to be applied to a Doc

dmonad commented on src/utils/DeleteSet.js:
```diff
@@ -278,37 +278,13 @@ export const readAndApplyDeleteSet = (decoder, transaction, store) => {
-            break
-          }
-        }
+        applyDeleteItem(transaction, structs, { clock, len })
```
This may sound silly, but creating a function call for every deleted `Item` really hurts performance.

Yjs is able to handle millions of `Item` integrations because the "hot-path" is inlined. I realize that there is more code that could be inlined. 

Please don't create a separate function call for deleting items.

calibr replied:
Hm, actually, I always thought that modern JS engines do function inlining itselves, so I didn't worry about that too much. With your updated approach using encoding/decoding we won't need `applyDeleteItem`, so I will remove this function.

dmonad replied:
You are right. JavaScript engines do inline code on a predictive basis. In nodejs you can inspect how many functions "can" be inlined and how many functions are actually inlined on execution. It seems there is an inline-budget (because inlining does duplicate code and increases the size of the heap). So inlining functions beforehand, if possible, does improve performance. I just started a few weeks ago to work on inlining the hot-path myself to see some slight improvements of 10-20%. I seriously don't recommend doing that for all projects. But it made sense for Yjs.

dmonad commented on src/utils/Snapshot.js:
```diff
@@ -148,3 +153,124 @@ export const splitSnapshotAffectedStructs = (transaction, snapshot) => {
+ * @param {Snapshot} snapshot
+ * @return {Doc}
+ */
+export const createDocFromSnapshot = (originDoc, snapshot) => {
```
`Item`s depend on each other and must be applied in a very specific order. The function `resumeStructIntegration` handles the correct integration process. If an `Item` depends on another `Item`, the dependency is integrated first.

In multi-user scenarios, the below integration process will likely result in issues. 

Copying `Item`s might be the fastest approach. But I think first we should implement a simple working implementation that just encodes the snapshotted document and then calls `applyUpdate` to make use of the existing struct integration process.

This is how I would lay out the implementation:

* Starting with a snapshot `{ sv, ds }`.
* Split items to ensure that really just the state is copied that we are interested in: `sv.forEach((clock, client) => getItemCleanStart(transaction, ID(client, clock)))` (note: first check if item `ID(client, clock)` exists, otherwise getItemCleanStart might throw an error).
* Create V2 encoder (v2 is probably faster).
* Write all structs from [0 - clock) to encoder.
* Write `ds` to encoder.
* The encoded document is now a document update containing all updates from the state vector
* Apply encoded document to a fresh Y.Doc

dmonad commented on tests/snapshot.tests.js:
```diff
@@ -0,0 +1,110 @@
+import { createDocFromSnapshot, Doc, snapshot, YMap } from '../src/internals'
```
Thanks for creating tests! These are very valuable!

We should also generate tests that contain multiple users that create changes that depend on each other. E.g. user1 inserts text and syncs with user2. User2 inserts text (the inserted `Item` depends on the content from user1).

calibr replied:
@dmonad thanks, do you mean we need to create a document with these dependent changes and then restore it to a specific state from the past? By generating tests, do you mean that they should be automatically generated somehow(in tests I saw `applyRandomTests` probably used for tests generation)?

dmonad replied:
>  thanks, do you mean we need to create a document with these dependent changes and then restore it to a specific state from the past? 

Exactly!

> By generating tests, do you mean that they should be automatically generated somehow(in tests I saw applyRandomTests probably used for tests generation)?

You don't have to at first. But it does make sense to do it before we release this feature. A lot of issues can be found with randomized testing. 

Maybe we can adapt `applyRandomTests` to create a snapshot at some point. At the end of the test we restore a document and check if the old values can be recovered.

calibr commented:
@dmonad hey Kevin, I updated the PR to address your change requests. Now it uses encoding/decoding to restore a snapshot and I added a tests for restoring documents with depended changes
